drivers/include/matrix_keypad.h

```diff
@@ -0,0 +1,218 @@
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    drivers_matrix_keypad Matrix Keypad
+ * @ingroup     drivers_sensors
+ * @brief       Matrix keypad driver for row/column keypads
+ *
+ * @{
+ *
+ * @file
+ *
+ * @author      Koen Zandberg <koen@bergzand.net>
+ *
+ * This module implements a simple matrix keypad driver where keys are connected
+ * between GPIO columns and rows. It works best with diodes in series with the
+ * switches to prevent key ghosting, but it can be used without these diodes.
+ * @ref CONFIG_MATRIX_KEYPAD_ROWS_USE_OPEN_DRAIN can be enabled when the keypad
+ * doesn't use diodes in the switches.
+ *
+ * The keypad works by configuring the column GPIOs as input with pull-ups. Each
+ * row is configured as open drain with pull-up. One by one the rows are set to
+ * pull their output low. For each row the column GPIOs are read and the state
+ * is checked. When a key is pressed the column GPIO of that switch will read
+ * low as soon as the row it is on is pulled low.
+ *
+ * The debouncing algorithm is a pattern style debounce where the switch must be
+ * in one position for a number of samples, then a set of "don't care" samples
+ * and then in the other position for a number of samples. The samples in the
+ * middle allow for a period where the switch can be either low or high without
+ * affecting the transition. The exact pattern is determined by
+ * @ref CONFIG_MATRIX_KEYPAD_DEBOUNCE_PATTERN_BEGIN and
+ * @ref CONFIG_MATRIX_KEYPAD_DEBOUNCE_PATTERN_END. These are used as mask where
+ * the switch must be in a determined state. The bits where neither pattern is
+ * set is used as the "don't care" set of samples.
+ *
+ * A full scan of the matrix keypad is done via the @ref matrix_keypad_scan
+ * function. This function iterates over all rows and columns to update the
+ * stored history of every pin.
+ *
+ * When a state change is detected on a switch, the @ref matrix_keypad_cb_t
+ * callback is called with the row and column number together with the new state
+ * of the switch (pressed or not pressed).
+ */
+
+#ifndef MATRIX_KEYPAD_H
+#define MATRIX_KEYPAD_H
+
+#include <stdint.h>
+#include <stdbool.h>
+#include "periph/gpio.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Maximum number of rows
+ */
+#ifndef CONFIG_MATRIX_KEYPAD_NUM_ROWS
+#define CONFIG_MATRIX_KEYPAD_NUM_ROWS   2
+#endif
+
+/**
+ * @brief Maximum number of columns
+ */
+#ifndef CONFIG_MATRIX_KEYPAD_NUM_COLUMNS
+#define CONFIG_MATRIX_KEYPAD_NUM_COLUMNS   2
+#endif
+
+/**
+ * @brief Debounce pattern high to low bits
+ */
+#ifndef CONFIG_MATRIX_KEYPAD_DEBOUNCE_PATTERN_BEGIN
+#define CONFIG_MATRIX_KEYPAD_DEBOUNCE_PATTERN_BEGIN  0xC0
+#endif
+
+/**
+ * @brief Debounce pattern low to high bits
+ */
+#ifndef CONFIG_MATRIX_KEYPAD_DEBOUNCE_PATTERN_END
+#define CONFIG_MATRIX_KEYPAD_DEBOUNCE_PATTERN_END  0x7
+#endif
```

Comment:
How would one adjust these?

3 bits set -> 3 cycles of `matrix_keypad_scan()` needed to assert state?

Reply:
It's also in C23 - but yea afaik we can't use it yet.

I'd still like to have the number of set bits -> number of required on/off cycles documented just to rule out any ambiguity.

Reply:
I elaborated a bit on the requirements here and tried to remove the ambiguity. Let me know what you think